feat: added rtl direction to ConfigProvider and some components

saeedrahimi opened the pull request:
<!--
First of all, thank you for your contribution! 😄

New feature please send pull request to feature branch, and rest to master branch.
Pull request will be merged after one of collaborators approve.
Please makes sure that these form are filled before submitting your pull request, thank you!

[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]
-->

### 🤔 This is a ...

- [x] New feature
- [x] Bug fix
- [x] Site / document update
- [x] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
[#4051](https://github.com/ant-design/ant-design/issues/4051)


### 💡 Background and solution
As requirement of this [issue](https://github.com/ant-design/ant-design/issues/4051) components of ant-design doesn't support right to left direction, and those components are not useful in some languages like Persian(Farsi) or Arabic.

Solution: 
I started to develop Rtl style on ant-design and as @zombieJ  mentioned in issue i used ConfigProvider to set direction of component.

layoutDirection property added to ConfigProvider.
```
<ConfigProvider layoutDirection="rtl">
      <Cascader  options={options} placeholder="select an item"/>
</ConfigProvider>
```
in style files added rtl based styles to change direction of displayed items
```
&-picker-small &-picker-clear,
  &-picker-small &-picker-arrow {
    right: @control-padding-horizontal-sm;
    [dir='rtl'] &, .rtl & {
      right: auto;
      left: @control-padding-horizontal;
    }
  }
```
so when one of parent elements has dir=['rtl'] (provided by ConfigProvider) paddings , margins, floats and other properties flipped to display correctly in right-to-left directions.

![cascader](https://user-images.githubusercontent.com/12451743/62215987-2dcb4d80-b3bd-11e9-9bc2-42b45a1cdad6.JPG)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   Added Rtl direction on cascader        |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed


-----
[View rendered components/cascader/demo/right-to-left.md](https://github.com/saeedrahimi/ant-design/blob/cascader-rtl/components/cascader/demo/right-to-left.md)

saeedrahimi committed:
add RTL direction to cascader with configProvider

saeedrahimi committed:
add RTL direction to cascader with configProvider

saeedrahimi committed:
lint: remove unused import

saeedrahimi committed:
fix CircleCli error and change rtl demo's debug mode

saeedrahimi committed:
update test snapshots

saeedrahimi commented:
Looks snapshots issue resolved in #18036 and it's on master branch. but my changes are on feature branch.
what can i do to fix this pull request?

zombieJ commented:
We will release minor version this week. which will sync  branch soon.

saeedrahimi committed:
Merge branch 'feature' of https://github.com/ant-design/ant-design into cascader-rtl

saeedrahimi committed:
update new snapshots after merge

zombieJ commented:
How about directly `direction` with value `ltr` & `rtl` like CSS: https://developer.mozilla.org/zh-CN/docs/Web/CSS/direction

@saeedrahimi  @ztplz @afc163

saeedrahimi commented:
I'm already using it in [these changes ](https://github.com/ant-design/ant-design/blob/ab8b12376756908fd48207f33b96fdc7b58520ea/components/config-provider/index.tsx#L95)
but the problem is some components needs to know the direction of page/view/parent to change behavior. [example](https://github.com/ant-design/ant-design/blob/ab8b12376756908fd48207f33b96fdc7b58520ea/components/cascader/index.tsx#L576) 

I'm not a React developer and i don't know about it's features. but in angular with [bidi](https://material.angular.io/cdk/bidi/api) you can detect inherited direction (RTL/LTR) of a component. i couldn't find anything like bidi in React, so i added variable to ConfigProvider to detect the direction.
I've already added RTL styles to most of components of [ng-zorro](https://ng.ant.design) library, but [they told me](https://github.com/NG-ZORRO/ng-zorro-antd/issues/1762#issuecomment-516315930) it should be implemented in ant-design first.

zombieJ commented:
I mean change prop name from `layoutDirection` to `direction`.

saeedrahimi committed:
change layoutDirection prop to direction and update demo

saeedrahimi committed:
change value order of direction prop

saeedrahimi committed:
update demo to fix ci error

wendzhue commented:
兴奋地搓手手

saeedrahimi committed:
Merge branch 'feature' of https://github.com/ant-design/ant-design into cascader-rtl

saeedrahimi commented:
any progress?

zombieJ commented on components/cascader/demo/right-to-left.md:
```diff
@@ -0,0 +1,89 @@
+
+## en-US
+
+Cascader component supports right to left (rtl) direction if placed inside `ConfigProvider` with `direction='rtl'`
```
This should also doc in `ConfigProvider` page.

saeedrahimi replied:
`ConfigProvider` documentations updated.

saeedrahimi committed:
doc: update configprovider documentation

saeedrahimi committed:
Merge branch 'feature' into cascader-rtl

saeedrahimi commented:
@zombieJ 
`FAIL  components/steps/__tests__/demo.test.js`
it has CI errors in step components.

zombieJ commented on components/config-provider/index.en-US.md:
```diff
@@ -43,3 +43,4 @@ Some component use dynamic style to support wave effect. You can config `csp` pr
 | getPopupContainer | to set the container of the popup element. The default is to create a `div` element in `body`. | Function(triggerNode) | `() => document.body` | 3.11.0 |
 | locale | language package setting, you can find the packages in [antd/es/locale](http://unpkg.com/antd/es/locale/) | object | 3.21.0 |
 | prefixCls | set prefix class | string | ant | 3.12.0 |
+| direction | to set direction of all elements underneath itself, set `rtl` to view underneath components right-to-left styled | string: 'ltr', 'rtl', 'auto' | ltr | 3.21.5 |
```
* Move Cascader demo into ConfigProvider.
* `set direction of text. See [demo](#your demo link)`

zombieJ replied:
We can add more component demo here.

saeedrahimi replied:
all done.

zombieJ commented on components/config-provider/index.zh-CN.md:
```diff
@@ -44,3 +44,4 @@ return (
 | getPopupContainer | 弹出框（Select, Tooltip, Menu 等等）渲染父节点，默认渲染到 body 上。 | Function(triggerNode) | () => document.body | 3.11.0 |
 | locale | 语言包配置，语言包可到 [antd/es/locale](http://unpkg.com/antd/es/locale/) 目录下寻找 | object | - | 3.21.0 |
 | prefixCls | 设置统一样式前缀 | string | ant | 3.12.0 |
+| direction | 要设置自身下方所有元素的方向，请将 `rtl` 设置为从右到左的样式下方查看 | string: 'ltr', 'rtl', 'auto' | ltr | 3.21.5 |
```
`设置文本展示方向。[示例](...)`

saeedrahimi committed:
fix: update configprovider demo and resolve conversation

saeedrahimi committed:
Update direction.md

saeedrahimi committed:
Update direction.md

zombieJ commented on components/config-provider/index.zh-CN.md:
```diff
@@ -44,4 +44,4 @@ return (
 | locale | 语言包配置，语言包可到 [antd/es/locale](http://unpkg.com/antd/es/locale/) 目录下寻找 | object | - | 3.21.0 |
 | prefixCls | 设置统一样式前缀 | string | ant | 3.12.0 |
-| direction | 要设置自身下方所有元素的方向，请将 `rtl` 设置为从右到左的样式下方查看 | string: 'ltr', 'rtl', 'auto' | ltr | 3.21.5 |
+| direction | 设置文本展示方向。[示例](/components/config-provider/#components-config-provider-demo-direction) | string: 'ltr', 'rtl', 'auto' | ltr | 3.21.5 |
```
`#components-config-provider-demo-direction` is good enought : )

zombieJ replied:
version: `3.21.5` -> `3.22.0`

saeedrahimi committed:
fix: resolve conversation issues.

saeedrahimi commented:
@zombieJ could you help me on resolving merge issue. it's about code coverage.

zombieJ commented on components/config-provider/demo/direction.md:
```diff
@@ -0,0 +1,172 @@
+            placeholder="Select an item"
+            popupPlacement={this.props.popupPlacement}
+            showSearch={this.cascaderFilter}
+          />
```
Could we provides more demo instead of Cascader only?

saeedrahimi replied:
yes, but for now i just changed styles only for Cascader and each component needs their rtl styles.
I've planned to change each components in separate pull request but I can change more components in this pull request if you want. :)

afc163 commented on components/config-provider/demo/direction.md:
```diff
@@ -0,0 +1,172 @@
+  render() {
+    return (
+      <div className="direction-components">
+        <div className="example">
```
Duplicated div for demo wrapper.

saeedrahimi replied:
resolved in commits

afc163 commented on components/config-provider/demo/direction.md:
```diff
@@ -0,0 +1,172 @@
+      direction: 'ltr',
+      popupPlacement: 'bottomLeft',
+    };
+  }
```
```suggestion
state = {
  direction: 'ltr',
  popupPlacement: 'bottomLeft',
};
```

afc163 commented on components/config-provider/index.tsx:
```diff
@@ -91,11 +96,13 @@ class ConfigProvider extends React.Component<ConfigProviderProps> {
-          {children}
-        </LocaleProvider>
-      </ConfigContext.Provider>
+      <div dir={direction}>
```
Should we add extra div to ConfigProvider?

saeedrahimi replied:
I couldn't find any proper element to add `dir` attribute to it, so i added this div.
Do you have any suggestion.

saeedrahimi replied:
And something i forgot to say is that i use `dir` attribute in less files to change style of component based on their direction.

afc163 replied:
Try pass `dir` in React context to every component.

saeedrahimi replied:
![screenshots](https://user-images.githubusercontent.com/12451743/63220452-49946900-c19d-11e9-9d8e-3d79f91ec1ad.png)

or think about a form, if it wrapped with `<div dir="rtl" />` all components get `dir` value from context except those are not React component.
![screenshots2](https://user-images.githubusercontent.com/12451743/63220520-c1af5e80-c19e-11e9-843e-a74f760f9faf.png)

saeedrahimi replied:
another approach is like your suggestion just use `dir` in components and let the user to design their forms or group of elements with rtl direction.

saeedrahimi replied:
like [this pen ](https://codepen.io/saeedr/pen/YzKWLQL) i've just added `dir="rtl` ro `Form` element and everything is ok.

saeedrahimi replied:
Removed extra div on `ConfigProvider`.

afc163 commented on site/theme/template/Content/ComponentDoc.jsx:
```diff
@@ -52,7 +52,13 @@ class ComponentDoc extends React.Component {
+      intl: { locale },
+      utils,
+      demos,
+    } = this.props;
```
Mixed some unrelated changes

saeedrahimi replied:
I can't exclude these files.
I ran this command and fetched upstream file.
`git checkout upstream/feature -- site/theme/template/Content/Demo.jsx`
but after commit and push files won't disappear from changes.

saeedrahimi replied:
Just fixed issue by editing file from browser and Github web. :|
I think on my local system it auto-format the file after commit and re-arranges the render method parameters.

saeedrahimi committed:
fix: remove duplicated div from demo

saeedrahimi committed:
Update components/config-provider/demo/direction.md

Co-Authored-By: 偏右 <afc163@gmail.com>

saeedrahimi committed:
Merge branch 'cascader-rtl' of https://github.com/saeedrahimi/ant-design into cascader-rtl

saeedrahimi committed:
fix: edit demo code to suggestion

saeedrahimi committed:
fix: edit demo code to suggestion

saeedrahimi committed:
revert changes to demo.jsx

saeedrahimi committed:
revert changes to demo.jsx

saeedrahimi committed:
revert changes to MainContent.jsx

saeedrahimi committed:
Reverse change to Article.jsx

saeedrahimi committed:
Revert changes to ComponentDoc.jsx

saeedrahimi committed:
Revert changes to Header.jsx

saeedrahimi committed:
Revert changes to index.jsx

saeedrahimi committed:
Revert changes to i18n.zh-CN.md

saeedrahimi committed:
Revert changes to mentions/index.tsx

saeedrahimi committed:
Merge branch 'feature' into cascader-rtl

saeedrahimi committed:
fix: remove configprovider wrapper div

saeedrahimi committed:
update test snapshots

saeedrahimi committed:
fix: direction.md eslint error

saeedrahimi commented:
@afc163
@zombieJ
any updates?

zombieJ commented:
hi @saeedrahimi, need more component in the demo of ConfigProvider:
> We can add more component demo here.

afc163 commented on docs/react/i18n.zh-CN.md:
```diff
@@ -65,7 +65,6 @@ return (
 | 泰米尔语         | ta_IN  |
 | 泰语             | th_TH  |
 | 土耳其语         | tr_TR  |
-| 罗马尼亚语       | ro_RO  |
```
Why remove this?

afc163 commented on components/style/core/rtl.less:
```diff
@@ -0,0 +1,8 @@
+}
+.ltr {
+  direction: ltr;
+}
```
Do we need global styles now?

saeedrahimi replied:
Maybe not, i'll remove it.

saeedrahimi replied:
These styles will be needed on some cases like direction of `.ant-col` that i'm used.
so every col wrapped in `.rtl` will be displayed correctly. (Used in ConfigProvider RTL demo)

saeedrahimi replied:
@afc163 , I removed `rtl` global styles and added prefixed class to every component. like `ant-cascader-picker-rtl`
Also i fixed grid system direction with another approach.

saeedrahimi committed:
Merge branch 'feature' into cascader-rtl

saeedrahimi committed:
revert i18n file changes

saeedrahimi committed:
added rtl styles for Tree component

saeedrahimi committed:
added rtl styles and demo for Switch

saeedrahimi committed:
added rtl styles to input and input group

saeedrahimi committed:
fixed Input.Search rtl styles

saeedrahimi committed:
added rtl direction and demo for select component

saeedrahimi committed:
remove workflow for first push

saeedrahimi committed:
removed workflow to allow push

saeedrahimi committed:
fix eslint errors

saeedrahimi commented:
> hi @saeedrahimi, need more component in the demo of ConfigProvider:
> 
> > We can add more component demo here.

I've edited more components and make them support RTL direction and updated ConfigProvider demo with more rtl components.

saeedrahimi commented on components/style/core/rtl.less:
```diff
@@ -0,0 +1,11 @@
+}
+.rtl .ant-col {
+  float: right;
+}
```
> Do we need global styles now?

These styles will be needed on some cases like direction of `.ant-col` that i'm used.
so every col wrapped in `.rtl` will be displayed correctly. (Used in ConfigProvider RTL demo)

zombieJ commented:
This can be adjust:
<img width="856" alt="屏幕快照 2019-09-25 上午10 41 01" src="https://user-images.githubusercontent.com/5378891/65564929-076efc00-df81-11e9-8199-a9dc71e94273.png">

saeedrahimi committed:
fix select.selectedValue adjustment

zombieJ commented:
Test failed...

saeedrahimi committed:
Merge branch 'feature' of https://github.com/ant-design/ant-design into feature

saeedrahimi committed:
Merge branch 'feature' into cascader-rtl

saeedrahimi committed:
lint styles

saeedrahimi commented:
>@zombieJ: Test failed...
Done :)

DeveloperFromUkraine commented:
Hither, what about Tabs ??? Do they support RTL ?

zombieJ commented:
> Hither, what about Tabs ??? Do they support RTL ?

@saeedrahimi, could we add Tabs in Demo also?

DeveloperFromUkraine commented:
> > Hither, what about Tabs ??? Do they support RTL ?
> 
> @saeedrahimi, could we add Tabs in Demo also?

Yeah, please if what I talk about this `::after` element of active Tab.

saeedrahimi commented:
> > Hither, what about Tabs ??? Do they support RTL ?
> 
> @saeedrahimi, could we add Tabs in Demo also?

Tabs is part of [reac-components](https://github.com/react-component/tabs) project and it should supported in that repository.
although i've created a [pull request](https://github.com/react-component/tabs/pull/227) on rc-tabs repository and implemented rtl support on it.

in ant-design some components need to change from their repository to support rtl.

issamYahiaoui reviewed (dismissed):
Works fine for me

MRezaSafari reviewed (dismissed):
Well Done.

saeedrahimi committed:
merge with upstream

shakesBeardZ commented:
@zombieJ can you tell us when you plan to merge this branch and ship it?

zombieJ committed:
fix: Add missing typescript define (#19251)

* fix: Add missing typescript define

* remove withStore

* fix lint

shaodahong committed:
refactor: improve typescript module declare (#19254)

* refactor: improve typescript declare

* remove rc-progress  declare

* remove rc-drawer  declare

* improve site antd module resolve path

hsuanxyz committed:
docs: fix page header API (#19257)

zombieJ committed:
fix: should work in old version react (#19262)

* fix: should work in old version react

* update ci config

* add version log

zombieJ committed:
docs: Update 3.24.1 changelog (#19270)

rayronvictor committed:
Fix: long author name breaks on small screen (#19272)

* Fix: long author name breaks on small screen

* Update index.less

qq645381995 committed:
fix: 修复点击下载按钮会执行预览方法 (#19268)

* fix: 修复点击下载按钮会执行预览方法

* chore: test

* chore: test

Commit:
Improve types for customRequest prop on Upload component

broder committed:
Fix typos in en-US contributing guide (#19277)

zombieJ committed:
chore: Remove React function polyfill (#19279)

* chore: Remove polyfill

* add more skip

shaodahong committed:
chore: fix compile fail when use cnpm installed

afc163 committed:
:bug: Fix Table scroll.x max-content not working

wleven committed:
Fix Select Mode Type

wleven committed:
Fix Select Mode Type

wleven committed:
Fix Select Mode Type

Niyiojeyinka committed:
corrected the typographical error of word 'case' to 'care' (#19290)

fix typo

afc163 committed:
:memo: :clapper: improve grid demo and documentaion

afc163 committed:
:sparkles: release 3.24.2

ycjcl868 committed:
test: official site checker before publish (#19247)

* tweak: official site checker before publish

* feat: check site files

* fix: diff logic

* fix: replace

* feat: add site jest test

* fix: test case

* fix: test

* feat: static server test

* Update check-site.js

* feat: en-US index

* fix: lint

neung123 committed:
fix(#18280):  add type [Moment.moment, undefined] to RangePickerValue (#19300)

zombieJ committed:
test: add test case for ref test (#19302)

* add test case for ref test

* update test case

afc163 committed:
:memo: fix wrong changelog

afc163 committed:
:memo: update changelog

afc163 committed:
Merge pull request #19304 from ant-design/master

chore: Merge master into feature

afc163 committed:
:lipstick: format getGutter code style

afc163 committed:
:bug: fix wrong responsive gutter

close #19307

afc163 committed:
:white_check_mark: add test cases

afc163 committed:
:memo: fix home page animation

0xPoe committed:
add commit lint action description

MRezaSafari commented:
@zombieJ any idea about this PR? is it gonna merge anytime soon?

afc163 committed:
:bug: fix small table border radius issue

close #19287

afc163 committed:
:bug: fix Menu in Dropdown submenu selected style

close #19264

zombieJ committed:
fix: Tree shaking should works on withContext

saeedrahimi committed:
Merge branch 'feature' into cascader-rtl

saeedrahimi committed:
remove unnecessary checked out files

afc163 committed:
:white_check_mark: update table snapshot

afc163 committed:
Merge pull request #19321 from ant-design/master

chore: sync master to feature

saeedrahimi committed:
update cascader tests

saeedrahimi committed:
merge with upstream/feature

saeedrahimi committed:
fix test lint errors

SaebaRyoo committed:
fix(#11119): fix Transfer listStyle support function callback (#19330)

* fix(#11119): fix Transfer listStyle support function callback

* fix(#11119): fix Transfer listStyle support function callback

saeedrahimi committed:
changed tree component global rtl style to local style

saeedrahimi committed:
changed cascader rtl class to prefixed class

saeedrahimi committed:
changed input rtl class to prefixed class

saeedrahimi committed:
fix cascader input text align on rtl

saeedrahimi committed:
make grid system compatible with RTL direction

DemetriusHR committed:
feat: Add tag downloadFile in all languages (#19361)

saeedrahimi committed:
fix cascader test fail

saeedrahimi committed:
removed global rtl style

afc163 commented:
It is better to add this feature in 4.0, could you change target branch to `4.0-prepare`?

saeedrahimi committed:
Merge remote-tracking branch 'upstream/feature' into cascader-rtl

saeedrahimi commented:
> It is better to add this feature in 4.0, could you change target branch to `4.0-prepare`?

target branch changed to  `4.0-prepare`

afc163 commented:
You should check out from `4.0-prepare`.

saeedrahimi commented:
> You should check out from `4.0-prepare`.

and open a new pull request?

afc163 commented:
> and open a new pull request?

Righto

saeedrahimi commented:
Moved to a new pull request #19380
